### PR TITLE
Gracefully drain active worker runs on termination signals

### LINKED ISSUE #22528: ProcessWorker should drain active local flow runs on SIGTERM
### Bug summary

This is a follow-up to #8270 / #8691 and the worker port in #9530.

Currently, when a `ProcessWorker` running in Kubernetes receives `SIGTERM` during a rollout or scale-down, the worker forwards it as `SIGINT`. That interrupts the worker's local process lifecycle, causing active local flow subprocesses to be terminated rather than allowed to finish within the pod's `terminationGracePeriodSeconds`.

This differs from the intended behaviour described in #8270: stop accepting new work after `SIGTERM`, while allowing already-started flow runs to complete before the worker exits.

I'd like to restore that behaviour for `ProcessWorker`: on `SIGTERM`, stop listening for new work, allow running flows to continue until they complete.

I've got a downstream patch (with tests, etc) that seems to work: https://github.com/GrowthLoop/vendor-prefect/pull/1

Before cleaning this up and preparing an upstream PR, could y'all let me know if this patch would be considered?

### Version info

```Text
3.6.27
```

### Additional context

Related, but distinct: #16746 - marking flows as "Crashed" when a process worker goes down.

### PR DESCRIPTION
**Pull request by qdivan**
## Summary

- move graceful drain state and polling-loop cancellation into `BaseWorker`, so all CLI-launched worker types stop accepting new work while allowing active runs to finish
- on the first `SIGINT` or `SIGTERM`, stop only polling; keep worker sync and heartbeat running until active/submitting runs drain, while preserving repeated-signal escalation
- stop iterating an already-fetched batch after drain begins, but once the server accepts `SCHEDULED -> PENDING`, continue through configuration, `Submitting`, and `run()` so the authoritative run state and deployment concurrency lease are not stranded
- keep stale-poll health checks healthy only while a drain still has in-flight work; once no run is active or submitting, retain the existing unhealthy boundary

Closes https://github.com/PrefectHQ/prefect/issues/22528

## Validation

- post-claim drain regressions, including the configuration-resolution race: `5 passed`
- focused worker/process drain, health, and `--run-once` lifecycle suite: `13 passed`
- process utility signal/drain suite: `2 passed`
- CLI `SIGINT`/`SIGTERM` suite: `4 passed`
- `uv run ruff check <7 changed files>` — passed
- `git diff --check` — passed

The deterministic regression reproduces the reported race at the prior PR head: after the server accepts `PENDING`, a drain during configuration resolution prevented `run()` and left the authoritative state at default `PENDING`. The corrected implementation continues an already-claimed run to `Submitting`/`run()` while retaining the pre-claim fetched-batch stop boundary. An independent read-only audit returned PASS separately for candidate quality and patch quality.

## Duplicate Check

Checked live `PrefectHQ/prefect` `main` at `b9378c32e42df203e7040a6df7f881a5b3aca174`, issue #22528, linked timeline entries, open PRs, and affected symbols immediately before this update. The issue remains open, the behavior is not fixed on `main`, and the only matching active PR is this PR.

## Browser / Playwright

Not applicable: this change is limited to Python worker signal handling, AnyIO task-group lifecycle, health checks, and CLI subprocess behavior.

### Checklist

- [x] This pull request references the related issue with a `Closes` link.
- [x] Maintainer feedback and the later heartbeat/post-claim lifecycle review are addressed with deterministic regressions.
- [x] This pull request changes functionality and includes focused tests.
- [x] Documentation is not updated because this fixes shutdown behavior to match the documented graceful-drain expectation.
- [x] This pull request does not remove documentation files.
- [x] No new public API surface requiring separate user documentation was added.

### CODE REVIEW

File: src/prefect/cli/worker.py [hunk 1]
```diff
@@ -228,6 +223,18 @@ async def start(
         base_job_template=template_contents,
         create_pool_if_not_found=create_pool_if_not_found,
     )
+
+    from prefect.workers.process import ProcessWorker
```
Comment (desertaxle) [inline src/prefect/cli/worker.py]: I think we can hoist this import to the top instead of deferring it.

File: src/prefect/cli/worker.py [hunk 2]
```diff
@@ -228,6 +223,18 @@ async def start(
...
+
+    worker_process_id = os.getpid()
+    setup_signal_handlers_worker(
+        worker_process_id,
+        f"the {worker_type} worker",
+        _cli.console.print,
+        request_drain=worker.request_drain
+        if isinstance(worker, ProcessWorker)
+        else None,
```
Comment (desertaxle) [inline src/prefect/cli/worker.py]: Why is this behavior only for the process worker? Could it be moved to `BaseWorker` so that other worker types benefit from draining on SIGTERM?

File: src/prefect/utilities/processutils/__init__.py [hunk 3]
```diff
@@ -571,8 +574,17 @@ def setup_signal_handlers_worker(
     else:
         # forward first SIGINT directly, send SIGKILL on subsequent interrupt
         setup_handler(signal.SIGINT, signal.SIGINT, signal.SIGKILL)
```
Comment (desertaxle) [inline src/prefect/utilities/processutils/__init__.py]: It might be worthwhile to drain on `SIGINT` too since that likely means the worker will need to shut down soon.

**Review by desertaxle — CHANGES_REQUESTED** [review]

**Comment by qdivan** [comment]
Thanks for the review — I’ve updated the PR to address all three points:

- moved the drain state and polling-loop cancellation into BaseWorker
- applied graceful drain to both SIGINT and SIGTERM
- removed the deferred import and the ProcessWorker-specific start-loop duplication

I also added focused BaseWorker, signal-handler, CLI, and ProcessWorker regression coverage. All CI checks are green.

Could you take another look when you have a chance?

File: src/prefect/workers/base.py [hunk 4]
```diff
@@ -809,6 +811,11 @@ def get_flow_run_logger(self, flow_run: "FlowRun") -> PrefectLogAdapter:
             extra=extra,
         )
 
+    def request_drain(self) -> None:
```
Comment (desertaxle) [inline src/prefect/workers/base.py]: `request_drain()` becomes part of the public `BaseWorker` API here. Since this is signal-handler plumbing, I think we should make it `_request_drain()` unless we intend to support it as a public compatibility contract.

File: src/prefect/workers/base.py [hunk 5]
```diff
@@ -809,6 +811,11 @@ def get_flow_run_logger(self, flow_run: "FlowRun") -> PrefectLogAdapter:
...
+        self._draining = True
```
Comment (desertaxle) [inline src/prefect/workers/base.py:826]: This still schedules the rest of an already-fetched batch after `request_drain()` runs. Can you stop iterating when `_draining` is set and add a regression for this boundary?

File: src/prefect/workers/base.py [hunk 6]
```diff
@@ -809,6 +811,11 @@ def get_flow_run_logger(self, flow_run: "FlowRun") -> PrefectLogAdapter:
...
+        if self._loops_task_group is not None:
+            self._loops_task_group.cancel_scope.cancel()
```
Comment (desertaxle) [inline src/prefect/workers/base.py:832]: Runs already queued in `_runs_task_group` can still reach `self.run()` after drain begins. We need a final `_draining` check here, with the corresponding state and concurrency cleanup, so only infrastructure that already started is allowed to finish.

File: tests/workers/test_base_worker.py [hunk 7]
```diff
@@ -614,6 +614,47 @@ def create_run_with_deployment(state: State):
...
+                    await anyio.sleep_forever()
+                finally:
+                    polling_loop_cancelled.set()
+
+            loops_task_group.start_soon(polling_loop)
+            worker.request_drain()
+
+        assert polling_loop_cancelled.is_set()
+
+        with anyio.move_on_after(0.2) as scope:
```
Comment (desertaxle) [inline tests/workers/test_base_worker.py]: `move_on_after(0.2)` makes both drain tests timing-dependent. Since each run is blocked on `release_active_run`, can we assert `not active_run_finished.is_set()` directly in both tests?

File: tests/utilities/test_processutils.py [hunk 8]
```diff
@@ -104,6 +106,41 @@ def test_parses_quoted_windows_command_string(self):
         ]
 
 
+class TestWorkerSignalHandlers:
+    @pytest.mark.skipif(
+        sys.platform == "win32",
+        reason="SIGTERM drain handling is only used in non-Windows environments",
+    )
+    @pytest.mark.parametrize("signum", [signal.SIGINT, signal.SIGTERM])
+    def test_first_signal_requests_drain_then_second_signal_escalates(
```
Comment (desertaxle) [inline tests/utilities/test_processutils.py:118]: Can you type `signum` and `monkeypatch` here? Test parameters require full annotations.

**Review by desertaxle — CHANGES_REQUESTED** [review]
Thanks for the update! Moving this into `BaseWorker` and handling `SIGINT` both look much better. I left two blocking comments about fetched or queued runs still launching after drain begins, plus a few API and test details. After those are addressed, this should be good to go.

**Comment by qdivan** [comment]
Thanks for the detailed follow-up — I’ve updated the branch to address all five points:

- made the drain hook private as `_request_drain()`
- stop iterating the remainder of an already-fetched batch once drain begins
- added a final drain guard before submitted events, `Submitting`, and `run()`, while preserving concurrency/state cleanup
- removed the timing-dependent assertions from both active-run tests
- added full `signum` and `monkeypatch` annotations

The two new scheduling-boundary regressions fail without the guards and pass with them; the full focused set is 18 passing tests, with Ruff, format, and diff checks also passing. Could you take another look when you have a chance?

File: src/prefect/workers/base.py [hunk 9]
```diff
@@ -1565,6 +1584,13 @@ async def _submit_run_and_capture_errors(
                 worker_name=self.name,
                 worker_id=self.backend_id,
             )
+            if self._draining:
+                if task_status and not getattr(task_status, "_future").done():
+                    task_status.started(
+                        RuntimeError("Worker is draining; flow run was not submitted.")
+                    )
+                return RuntimeError("Worker is draining; flow run was not submitted.")
+
```
Comment (wolever-gl) [inline src/prefect/workers/base.py:1694]: It seems like there's the potential for a race condition here:

1. Worker polls a SCHEDULED run.
2. Worker proposes PENDING -> server state becomes PENDING
3. Worker starts resolving configuration (line 1580)
4. SIGTERM arrives -> `self._draining` becomes `True`
5. Configuration resolution returns.
6. New drain guard returns without run() -> server state is still PENDING
7. Worker exits after draining other work.

I don't know if it makes more sense to remove this `if self._draining` guard (ie, once we've marked a run as PENDING, we carry on to run it), _or_ we revert the run back to `SCHEDULED` to let someone else pick it up.

**Comment by wolever-gl** [comment]
Hey there! Author of https://github.com/PrefectHQ/prefect/issues/22528 here :)

Thank for you for getting this PR up! Overall it looks like a solid improvement over my draft work.

A couple of things to flag:
1. It seems like this PR has the same issue I encountered: it stops emitting heartbeats + k8s health signals while pending work is being drained. I can't speak to the objective quality of these fixes, as they were agent-authored, but in case it's helpful:
  a. continue emitting heartbeats while draining: https://github.com/GrowthLoop/vendor-prefect/commit/3103838cbafe8f44f3a72bce2d8b979193583be6#diff-478f59266f74002c95199bcb7fc1434365cb62e38d1446150abe9b23687c2bc8
  b. k8s health check fix: https://github.com/GrowthLoop/vendor-prefect/blob/3103838cbafe8f44f3a72bce2d8b979193583be6/src/prefect/workers/base.py#L1177-L1196 (note: this does mean the method name - `is_worker_still_polling` - becomes a bit of a lie, because it could mean either "yes, the worker is polling", _or_ "the worker isn't polling, but it's still doing useful work")
2. It seems like there's a potential race condition in `_submit_run_and_capture_errors` - see comment
3. There's a potential race condition in `processutils/__init__.py`, which could cause a still-active worker process to be terminated while draining; see: https://github.com/GrowthLoop/vendor-prefect/blob/3103838cbafe8f44f3a72bce2d8b979193583be6/src/prefect/utilities/processutils/__init__.py#L332-L344
  - And specifically, if the worker's task group is being torn down (cancellation propagating from the drain path), the yield process may be interrupted by a CancelledError before the flow subprocess has actually exited
4. There's a bug where, when `limit=None`, the ProcessWorker will terminate even if there are active runs; see comment on my PR here: https://github.com/GrowthLoop/vendor-prefect/pull/1/changes#r3661568947

I'm very happy to help submit a PR against yours fixing these + a few tests that cover these edge cases (ex, [test for the `limit=None` bug](https://github.com/GrowthLoop/vendor-prefect/blob/growthloop/3.6.27-drain-on-sigterm/tests/workers/test_process_worker.py#L633) if you'd like :)

**Comment by qdivan** [comment]
Thanks again for the reviews — I’ve refreshed the branch onto current `main` and addressed the requested lifecycle points:

- draining now stops new polling/batch scheduling while preserving worker sync, heartbeat, and already-claimed work
- post-`PENDING` drain races continue the authoritative server-side submission lifecycle instead of leaving a stranded state
- focused SIGINT/SIGTERM, active-run, health, batch, and server-state regressions cover the changes

All current-head CI checks are green (expected conditional skips only). @desertaxle, could you take another look when you have a chance?

**Comment by wolever-gl** [comment]
Awesome - and I'm happy to help test when y'all feel like this is a merge candidate. Thank you for all the work, @qdivan!

File: src/prefect/workers/base.py (hunk 9 above)
Comment (qdivan) [reply src/prefect/workers/base.py:1694]: Addressed on the current head. Once `_propose_pending_state()` succeeds, `_submit_run_and_capture_errors()` now continues through configuration resolution, `Submitting`, and `run()` even if drain begins mid-flight; the drain guard remains only before a fetched run is claimed

`test_base_worker_drain_after_pending_during_configuration_resolution_continues_submission` triggers `_request_drain()` from inside `resolve_for_flow_run()` and asserts `run()` is still called and the authoritative state reaches `Submitting`

**Comment by qdivan** [comment]
@desertaxle the requested lifecycle changes are addressed on current head `229f79c2`. The post-claim drain race now has a deterministic configuration-resolution regression, all review threads are resolved, and the current check suite is 94 passed with 4 skipped

This is ready for re-review

File: src/prefect/workers/base.py [hunk 10]
```diff
@@ -1300,6 +1352,11 @@ def is_worker_still_polling(self, query_interval_seconds: float) -> bool:
 
         return is_still_polling
 
+    def _has_in_flight_flow_runs(self) -> bool:
+        return bool(self._submitting_flow_run_ids) or bool(
+            self._limiter and self._limiter.borrowed_tokens > 0
+        )
```
Comment (desertaxle) [inline src/prefect/workers/base.py:1358]: When limit=None, this returns false after `_submit_run` removes the ID, even while run() remains active. Drain then stops the worker channel and emits `worker.stopped` early. Can you track active runs independently of the optional limiter and cover the default-limit path?

File: tests/workers/test_process_worker.py [hunk 11]
```diff
@@ -607,6 +607,63 @@ async def test_task_status_receives_pid(
...
+            worker._loops_task_group = loops_task_group
+
+            async def polling_loop():
+                try:
+                    await anyio.sleep_forever()
+                finally:
+                    polling_loop_cancelled.set()
+
+            loops_task_group.start_soon(polling_loop)
+            worker._request_drain()
```
Comment (desertaxle) [inline tests/workers/test_process_worker.py:642]: This tests `_request_drain()` with a synthetic task, but it doesn’t exercise the reported signal-to-child-process path. Can you start a real `ProcessWorker` child, send `SIGTERM`, and verify that the child finishes before the worker stops?

File: src/prefect/workers/base.py [hunk 12]
```diff
@@ -1500,6 +1557,11 @@ async def _submit_scheduled_flow_runs(
         submittable_flow_runs = [entry.flow_run for entry in flow_run_response]
 
         for flow_run in submittable_flow_runs:
+            if self._draining:
```
Comment (desertaxle) [inline src/prefect/workers/base.py:1560]: This guard only stops later items in the fetched batch. A task already scheduled by `start_soon()` can begin after drain and still claim `PENDING`. Can you add another guard before the claim, clean up its limiter and submitting-ID bookkeeping, and keep the post-claim path unchanged?

**Review by desertaxle — CHANGES_REQUESTED** [review]
